bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java

```diff
@@ -460,6 +471,12 @@ public HapiConverter<Schema> visitComposite(String elementName,
 
     if (converter == null) {
 
+      // We don't want 'id' to be present in nested fields to make it consistent with SQL-ON-FHIR.
+      // https://github.com/FHIR/sql-on-fhir/blob/master/sql-on-fhir.md#id-fields-omitted
+      if (!isRootElement) {
+        children.removeIf(field -> field.fieldName().equals("id"));
```

Comment:
I feel uneasy about this change because we are mutating an input argument. This is an inherent problem with Java that we cannot enforce input immutability (e.g., like C++ with _const_) but usually we expect the input arguments to a function not to be mutated by that function, unless it is clear from the name or is documented. So I treat the `children` input here as `cost List<>& children` (mimicking C++) and `removeIf` would not be a `const` method. If we drop that immutable assumption, the code quickly becomes hard to reason about.

So my suggestion is this: Remove the new `isRootElement` from this `visitComposite` method. Instead, deal with the `children` list in the caller method similar to how you have done this at line 597 of `R4StructureDefinitions.java` (in the `transform` method). There, we create the `children` list (so we are "the owner") and right away remove the `id` element, so calling `visitComposite` on that list works just fine. I understand this may need a few more changes (I think there are a few calls to `visitComposite` in each of R4 and STU3 versions) but I think that is a cleaner solution.

Reply:
Thanks, made changes as suggested.